### PR TITLE
Add readiness endpoint

### PR DESCRIPTION
**Pull request by lucacome**
Implement the readiness endpoint. The endpoint returns `503` while NGINX is starting up and loading all the configs. As soon as this is done the endpoint will return `200`.

There might be more docs needed.

### CODE REVIEW

File: deployments/deployment/nginx-ingress.yaml [hunk 1]
```diff
@@ -28,6 +28,13 @@ spec:
           containerPort: 443
        #- name: prometheus
          #containerPort: 9113
+       #- name: readiness-port
```
Comment (Rulox) [inline deployments/deployment/nginx-ingress.yaml:31]: We need to add these changes in `deployments/daemonset/nginx-ingress.yaml` and  `deployments/daemonset/nginx-plus-ingress.yaml` right?

File: cmd/nginx-ingress/main.go [hunk 2]
```diff
@@ -146,6 +146,10 @@ var (
 	spireAgentAddress = flag.String("spire-agent-address", "",
 		`Specifies the address of the running Spire agent. For use with NGINX Service Mesh only. If the flag is set,
 			but the Ingress Controller is not able to connect with the Spire Agent, the Ingress Controller will fail to start.`)
+
+	readyStatus = flag.Bool("ready-status", false, "Enable readiness endpoint '/nginx-ready'. The endpoint returns a success code when NGINX has loaded all the config after startup")
```
Comment (Rulox) [inline cmd/nginx-ingress/main.go:150]:
```suggestion
	readyStatus = flag.Bool("ready-status", false, "Enables readiness endpoint '/nginx-ready'. The endpoint returns a success code when NGINX has loaded all the config after startup")
```

File: deployments/helm-chart/README.md [hunk 3]
```diff
@@ -242,6 +242,8 @@ Parameter | Description | Default
 `controller.reportIngressStatus.leaderElectionLockName` | Specifies the name of the ConfigMap, within the same namespace as the controller, used as the lock for leader election. controller.reportIngressStatus.enableLeaderElection must be set to true. | Autogenerated
 `controller.reportIngressStatus.annotations` | The annotations of the leader election configmap. | {}
 `controller.pod.annotations` | The annotations of the Ingress Controller pod. | {}
+`controller.readyStatus.enable` | Enables liveness endpoint `"/nginx-ready"`. The endpoint returns a success code when NGINX has loaded all the config after startup. | false
```
Comment (Rulox) [inline deployments/helm-chart/README.md:245]:
```suggestion
`controller.readyStatus.enable` | Enables readiness endpoint `"/nginx-ready"`. The endpoint returns a success code when NGINX has loaded all the config after startup. | false
```

File: deployments/helm-chart/values.yaml [hunk 4]
```diff
@@ -235,6 +235,12 @@ controller:
 
   ## The PriorityClass of the ingress controller pods.
   priorityClassName:
+  readyStatus:
+    ## Enable readiness endpoint "/nginx-ready". The endpoint returns a success code when NGINX has loaded all the config after startup.
```
Comment (Rulox) [inline deployments/helm-chart/values.yaml:239]:
```suggestion
    ## Enables readiness endpoint "/nginx-ready". The endpoint returns a success code when NGINX has loaded all the config after startup.
```

File: docs-web/configuration/global-configuration/command-line-arguments.md [hunk 5]
```diff
@@ -176,4 +176,15 @@ Below we describe the available command-line arguments:
 	Specifies the address of a running Spire agent. **For use with NGINX Service Mesh only**.
 
 	- If the argument is set, but the Ingress Controller is unable to connect to the Spire Agent, the Ingress Controller will fail to start.
+
+.. option:: -ready-status
+
+	Enables liveness endpoint "/nginx-ready". The endpoint returns a success code when NGINX has loaded all the config after startup.
```
Comment (Rulox) [inline docs-web/configuration/global-configuration/command-line-arguments.md:182]:
```suggestion
	Enables readiness endpoint "/nginx-ready". The endpoint returns a success code when NGINX has loaded all the config after startup.
```

File: docs-web/configuration/global-configuration/command-line-arguments.md [hunk 6]
```diff
@@ -176,4 +176,15 @@ Below we describe the available command-line arguments:
...
+
+.. option:: -ready-status-port
+
+	The HTTP port for the liveness endpoint.
```
Comment (Rulox) [inline docs-web/configuration/global-configuration/command-line-arguments.md:186]:
```suggestion
	The HTTP port for the readiness endpoint.
```

File: deployments/helm-chart/README.md [hunk 7]
```diff
@@ -242,6 +242,8 @@ Parameter | Description | Default
...
+ `controller.readyStaus.port` | The HTTP port for the liveness endpoint | 8081
```
Comment (Rulox) [inline deployments/helm-chart/README.md:246]:
```suggestion
 `controller.readyStaus.port` | The HTTP port for the readiness endpoint | 8081
```

File: docs-web/configuration/global-configuration/command-line-arguments.md (hunk 6 above)
Comment (lucacome) [reply docs-web/configuration/global-configuration/command-line-arguments.md:186]: Thanks for catching these, not sure why I started thinking about liveness lol

File: deployments/daemon-set/nginx-ingress.yaml [hunk 8]
```diff
@@ -29,6 +29,13 @@ spec:
           hostPort: 443
        #- name: prometheus
          #containerPort: 9113
+       #- name: readiness-port
+         #containerPort: 8081
+        #readinessProbe:
```
Comment (pleshakov) [inline deployments/daemon-set/nginx-ingress.yaml:34]: we comment out the fields in the YAML by adding `#` in front of the name of a field. Note that the first letter of the field is in its correct position.  The readinessProbe field doesn't follow that convention. Could you fix that?

(before fixing, please see the suggestion about enabling readiness by default)

File: deployments/helm-chart/README.md [hunk 9]
```diff
@@ -242,6 +242,8 @@ Parameter | Description | Default
 `controller.reportIngressStatus.leaderElectionLockName` | Specifies the name of the ConfigMap, within the same namespace as the controller, used as the lock for leader election. controller.reportIngressStatus.enableLeaderElection must be set to true. | Autogenerated
 `controller.reportIngressStatus.annotations` | The annotations of the leader election configmap. | {}
 `controller.pod.annotations` | The annotations of the Ingress Controller pod. | {}
+`controller.readyStatus.enable` | Enables readiness endpoint `"/nginx-ready"`. The endpoint returns a success code when NGINX has loaded all the config after startup. | false
```
Comment (pleshakov) [inline deployments/helm-chart/README.md:245]:
```suggestion
`controller.readyStatus.enable` | Enables the readiness endpoint `"/nginx-ready"`. The endpoint returns a success code when NGINX has loaded all the config after the startup. This also configures a readiness probe for the Ingress Controller pods that uses the readiness endpoint. | false
```

File: deployments/helm-chart/README.md [hunk 10]
```diff
@@ -242,6 +242,8 @@ Parameter | Description | Default
...
+ `controller.readyStaus.port` | The HTTP port for the readiness endpoint | 8081
```
Comment (pleshakov) [inline deployments/helm-chart/README.md:246]:
```suggestion
 `controller.readyStaus.port` | The HTTP port for the readiness endpoint. | 8081
```

File: cmd/nginx-ingress/main.go [hunk 11]
```diff
@@ -146,6 +146,10 @@ var (
 	spireAgentAddress = flag.String("spire-agent-address", "",
 		`Specifies the address of the running Spire agent. For use with NGINX Service Mesh only. If the flag is set,
 			but the Ingress Controller is not able to connect with the Spire Agent, the Ingress Controller will fail to start.`)
+
+	readyStatus = flag.Bool("ready-status", false, "Enables readiness endpoint '/nginx-ready'. The endpoint returns a success code when NGINX has loaded all the config after startup")
```
Comment (pleshakov) [inline cmd/nginx-ingress/main.go:150]:
```suggestion
	readyStatus = flag.Bool("ready-status", false, "Enables the readiness endpoint '/nginx-ready'. The endpoint returns a success code when NGINX has loaded all the config after the startup")
```

File: deployments/helm-chart/templates/controller-deployment.yaml [hunk 12]
```diff
@@ -130,4 +139,8 @@ spec:
           - -enable-tls-passthrough={{ .Values.controller.enableTLSPassthrough }}
           - -enable-snippets={{ .Values.controller.enableSnippets }}
 {{- end }}
+{{- if .Values.controller.readyStatus.enable }}
```
Comment (pleshakov) [inline deployments/helm-chart/templates/controller-deployment.yaml:142]: it is also possible to simplify if you wish, like in the example below:
```
          - -health-status={{ .Values.controller.healthStatus }}
          - -health-status-uri={{ .Values.controller.healthStatusURI }}
```

File: deployments/helm-chart/values.yaml [hunk 13]
```diff
@@ -235,6 +235,12 @@ controller:
 
   ## The PriorityClass of the ingress controller pods.
   priorityClassName:
+  readyStatus:
```
Comment (pleshakov) [inline deployments/helm-chart/values.yaml:238]: we have an empty line between parameters

File: internal/k8s/controller.go [hunk 14]
```diff
@@ -743,6 +746,12 @@ func (lbc *LoadBalancerController) sync(task task) {
 	case transportserver:
 		lbc.syncTransportServer(task)
 	}
+
+	if lbc.firstRun && lbc.syncQueue.Len() == 0 {
```
Comment (pleshakov) [inline internal/k8s/controller.go:750]: I wonder if we need to have the `firstRun` field at all. What if we had just this:
```go
	if !lbc.isNginxReady && lbc.syncQueue.Len() == 0 {
		lbc.isNginxReady = true
		glog.V(3).Infof("NGINX is ready")
	}
```

File: internal/k8s/task_queue.go [hunk 15]
```diff
@@ -64,6 +64,11 @@ func (tq *taskQueue) Requeue(task task, err error) {
 	tq.queue.Add(task)
 }
 
+func (tq *taskQueue) Len() int {
```
Comment (pleshakov) [inline internal/k8s/task_queue.go:67]: missing a doc string

File: cmd/nginx-ingress/main.go [hunk 16]
```diff
@@ -631,3 +647,13 @@ func validateLocation(location string) error {
 	}
 	return nil
 }
+
+func ready(lbc *k8s.LoadBalancerController) http.HandlerFunc {
+	return func(w http.ResponseWriter, _ *http.Request) {
+		if !lbc.IsNginxReady() {
```
[the remark this answers is not in the document]
Comment (lucacome) [reply cmd/nginx-ingress/main.go:734]: I've added a response for when it's ready, it's already saying that the service is unavailable when it's not.

File: internal/k8s/controller.go (hunk 14 above)
Comment (lucacome) [reply internal/k8s/controller.go]: I think we need it to know if the flag is enabled or not.

File: docs-web/configuration/global-configuration/command-line-arguments.md [hunk 17]
```diff
@@ -188,5 +188,15 @@ Below we describe the available command-line arguments:
    Requires :option:`-nginx-plus`
 
 	 - If the argument is set, but `nginx-plus` is set to false, the Ingress Controller will fail to start.
-	 
+
+.. option:: -ready-status
+
+	Enables readiness endpoint "/nginx-ready". The endpoint returns a success code when NGINX has loaded all the config after startup.
```
Comment (pleshakov) [inline docs-web/configuration/global-configuration/command-line-arguments.md:194]:
```suggestion
 	Enables the readiness endpoint "/nginx-ready". The endpoint returns a success code when NGINX has loaded all the config after the startup.
```

File: internal/k8s/controller.go (hunk 14 above)
Comment (pleshakov) [reply internal/k8s/controller.go]: @lucacome 
is there any harm if we do that check regardless of whether `firstRun` enabled or not?

**Pull request merged**